DMTN-229.tex

```diff
@@ -1,39 +1,435 @@
+The Quantum Graph itself provides provenance information and is tied to the datasets by preserving the predicted output dataset UUIDs in the Butler when they are stored during processing.
+
+\begin{figure}
+  (a) \includegraphics[width=0.95\textwidth]{pipeline}
+  (b) \includegraphics[width=0.95\textwidth]{quantum-graph}
+  \caption{Top is a visualization of the pipeline shown in Fig.~\ref{fig:pipeline}. Bottom is the Quantum Graph for a single frame processing of one detector from one HSC observation showing the same pipeline with specific datasets.}
+  \label{fig:graph}
+  \end{figure}
+
+\section{Integration with Batch Systems}
+
+A Quantum Graph describes the work that the pipeline has to do in order to process all the data, but it does not say how the work should be scheduled or any resources that are required.
+To run a Quantum Graph it has to be converted to something that can be executed.
+On a single compute node (including a laptop) the \texttt{ctrl\_mpexec} package \footnote{\url{https://github.com/lsst/ctrl_mpexec}} provides a multi-processing executor, commonly run via the \texttt{pipetask} command line, that can run every quantum in the correct order and track all the responses.
+
+When a Quantum Graph consists of thousands of quanta, a single compute node is no longer sufficient.
+At that scale we need to use batch computing with tooling such as HTCondor,\cite{10.1002/cpe.938} Pegasus,\cite{10.1016/j.future.2014.10.008} Parsl,\cite{10.1145/3307681.3325400} or PanDA.\cite{10.1088/1742-6596/331/7/072024}
+For these batch systems the Quantum Graph must be translated into the form expected by the target system.
+A prototype demonstrated the approach on AWS with Pegasus using an S3 object store\cite{2020arXiv201106044B} but that used custom code.
+It was soon realized that many of the concepts for translating a Quantum Graph to a specialist graph are common and we developed a new Batch Production System (\texttt{ctrl\_bps}\footnote{\url{https://github.com/ctrl_bps}}) that would provide a framework to allow the aforementioned batch environments to be treated as plugin code.
+
+The BPS system provides a facility for a Quantum Graph to be converted to a generic workflow graph before being converted to the final form.
+This intermediate graph can take configuration parameters that specify special resource requirements (CPU or memory) required by specific tasks.
+It also allows quanta to be clustered for efficiency.
+Some quanta can be quick, executing in tens of seconds, and not all batch systems are designed to efficiently handle such small payloads.
+Clustering allows for multiple quanta to be combined into a single execution job, this can be particularly efficient if the clustering is designed such that the output of the first quantum is the only input to the next quantum (which is a common scenario for the early stages of many pipelines when a single detector is being characterized).
```

Comment:
```suggestion
Clustering allows for multiple quanta to be combined into a single execution job; this can be particularly efficient if the clustering is designed such that the output of the first quantum is the only input to the next quantum (which is a common scenario for the early stages of many pipelines when a single detector is being characterized).
```